File: packages/go/edgesign/edgesign_test.go
package edgesign

import (
	"encoding/json"
	"errors"
	"net/url"
	"os"
	"strings"
	"testing"
	"time"
)

type vector struct {
	Name  string `json:"name"`
	Input struct {
		Method                     string `json:"method"`
		BaseURL                    string `json:"baseUrl"`
		Bucket                     string `json:"bucket"`
		Key                        string `json:"key"`
		Secret                     string `json:"secret"`
		Expires                    int64  `json:"expires"`
		Range                      string `json:"range"`
		ResponseContentType        string `json:"responseContentType"`
		ResponseContentDisposition string `json:"responseContentDisposition"`
	} `json:"input"`
	Now         int64  `json:"now"`
	Canonical   string `json:"canonical"`
	URL         string `json:"url"`
	VerifyRange string `json:"verifyRange"`
}

func TestVectors(t *testing.T) {
	for _, vector := range loadVectors(t) {
		t.Run(vector.Name, func(t *testing.T) {
			expires := time.Unix(vector.Input.Expires, 0).UTC()
			raw, err := SignURL(SignInput{
				Method:                     vector.Input.Method,
				BaseURL:                    vector.Input.BaseURL,
				Bucket:                     vector.Input.Bucket,
				Key:                        vector.Input.Key,
				Secret:                     vector.Input.Secret,
				Expires:                    expires,
				Range:                      vector.Input.Range,
				ResponseContentType:        vector.Input.ResponseContentType,
				ResponseContentDisposition: vector.Input.ResponseContentDisposition,
			})
			if err != nil {
				t.Fatalf("SignURL() error = %v", err)
			}
			if raw != vector.URL {
				t.Fatalf("SignURL() = %q, want %q", raw, vector.URL)
			}

			claims, err := VerifyURL(VerifyInput{Method: vector.Input.Method, URL: raw, Secret: vector.Input.Secret, Now: time.Unix(vector.Now, 0).UTC(), Range: vector.VerifyRange})
			if err != nil {
				t.Fatalf("VerifyURL() error = %v", err)
			}
			if got := CanonicalString(claims); got != vector.Canonical {
				t.Fatalf("CanonicalString() = %q, want %q", got, vector.Canonical)
			}
		})
	}
}

func TestVerifyRangeHeaderHandling(t *testing.T) {
	now := time.Unix(1780934400, 0).UTC()
	raw, err := SignURL(SignInput{Method: "GET", BaseURL: "https://files.example.com", Bucket: "demo-bucket", Key: "object.txt", Secret: "secret", Expires: now.Add(time.Minute)})
	if err != nil {
		t.Fatal(err)
	}
	_, err = VerifyURL(VerifyInput{Method: "GET", URL: raw, Secret: "secret", Now: now, Range: "bytes=0-9"})
	if !errors.Is(err, ErrUnsignedRange) {
		t.Fatalf("VerifyURL() error = %v, want ErrUnsignedRange", err)
	}

	ranged, err := SignURL(SignInput{Method: "GET", BaseURL: "https://files.example.com", Bucket: "demo-bucket", Key: "object.txt", Secret: "secret", Expires: now.Add(time.Minute), Range: "bytes=0-9"})
	if err != nil {
		t.Fatal(err)
	}
	_, err = VerifyURL(VerifyInput{Method: "GET", URL: ranged, Secret: "secret", Now: now, Range: "bytes=1-9"})
	if !errors.Is(err, ErrRangeMismatch) {
		t.Fatalf("VerifyURL() error = %v, want ErrRangeMismatch", err)
	}
}

func TestMultiServerSignAndVerify(t *testing.T) {
	now := time.Date(2026, 6, 8, 12, 0, 0, 0, time.UTC)
	expires := now.Add(time.Minute)
	raw, err := SignURL(SignInput{
		Method:                     "GET",
		BaseURL:                    "https://files.example.com",
		Server:                     "edge-1",
		Bucket:                     "demo-bucket",
		Key:                        "dir/object.txt",
		Secret:                     "top-secret",
		Expires:                    expires,
		Range:                      "bytes=0-9",
		ResponseContentType:        "text/plain",
		ResponseContentDisposition: "attachment; filename=object.txt",
	})
	if err != nil {
		t.Fatalf("SignURL() error = %v", err)
	}
	u, err := url.Parse(raw)
	if err != nil {
		t.Fatalf("parse signed URL: %v", err)
	}
	if got, want := u.EscapedPath(), "/edge-1/demo-bucket/dir/object.txt"; got != want {
		t.Fatalf("path = %q, want %q", got, want)
	}

	claims, err := VerifyURL(VerifyInput{Method: "GET", URL: raw, Server: "edge-1", Secret: "top-secret", Now: now, Range: "bytes=0-9"})
	if err != nil {
		t.Fatalf("VerifyURL() error = %v", err)
	}
	if claims.Server != "edge-1" || claims.Bucket != "demo-bucket" || claims.Key != "dir/object.txt" {
		t.Fatalf("claims = %#v", claims)
	}
	wantCanonical := "GET\nedge-1\ndemo-bucket\ndir/object.txt\n" + formatUnixSeconds(expires) + "\nbytes=0-9\ntext/plain\nattachment; filename=object.txt"
	if got := CanonicalString(claims); got != wantCanonical {
		t.Fatalf("CanonicalString() = %q, want %q", got, wantCanonical)
	}
}

func TestVerifyURLMultiServerRejectsTamperAndExpectedServerMismatch(t *testing.T) {
	now := time.Date(2026, 6, 8, 12, 0, 0, 0, time.UTC)
	raw, err := SignURL(SignInput{Method: "GET", BaseURL: "https://files.example.com", Server: "edge-1", Bucket: "demo-bucket", Key: "object.txt", Secret: "secret", Expires: now.Add(time.Minute)})
	if err != nil {
		t.Fatal(err)
	}

	tampered := strings.Replace(raw, "/edge-1/", "/edge-2/", 1)
	_, err = VerifyURL(VerifyInput{Method: "GET", URL: tampered, Server: "edge-1", Secret: "secret", Now: now})
	if !errors.Is(err, ErrInvalidSignature) {
		t.Fatalf("VerifyURL() tamper error = %v, want ErrInvalidSignature", err)
	}

	_, err = VerifyURL(VerifyInput{Method: "GET", URL: raw, Server: "edge-2", Secret: "secret", Now: now})
	if !errors.Is(err, ErrInvalidSignature) {
		t.Fatalf("VerifyURL() server mismatch error = %v, want ErrInvalidSignature", err)
	}
}

func TestNoServerUsesLegacySingleServerBehavior(t *testing.T) {
	now := time.Date(2026, 6, 8, 12, 0, 0, 0, time.UTC)
	expires := now.Add(time.Minute)
	raw, err := SignURL(SignInput{Method: "GET", BaseURL: "https://files.example.com", Bucket: "demo-bucket", Key: "object.txt", Secret: "secret", Expires: expires})
	if err != nil {
		t.Fatalf("SignURL() error = %v", err)
	}
	u, err := url.Parse(raw)
	if err != nil {
		t.Fatalf("parse signed URL: %v", err)
	}
	if got, want := u.EscapedPath(), "/demo-bucket/object.txt"; got != want {
		t.Fatalf("path = %q, want %q", got, want)
	}

	claims, err := VerifyURL(VerifyInput{Method: "GET", URL: raw, Secret: "secret", Now: now})
	if err != nil {
		t.Fatalf("VerifyURL() error = %v", err)
	}
	if claims.Server != "" {
		t.Fatalf("claims.Server = %q, want empty", claims.Server)
	}
	wantCanonical := "GET\ndemo-bucket\nobject.txt\n" + formatUnixSeconds(expires) + "\n\n\n"
	if got := CanonicalString(claims); got != wantCanonical {
		t.Fatalf("CanonicalString() = %q, want %q", got, wantCanonical)
	}
}

func loadVectors(t *testing.T) []vector {
	t.Helper()
	data, err := os.ReadFile("../../testdata/edge-signing-vectors.json")
	if err != nil {
		t.Fatal(err)
	}
	var vectors []vector
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatal(err)
	}
	return vectors
}

File: packages/ts/test/edgesign.test.ts
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';
import test from 'node:test';

import {
  canonicalString,
  EdgeSignError,
  InvalidSignatureError,
  RangeMismatchError,
  signUrl,
  UnsignedRangeError,
  verifyUrl,
} from '../src/index.ts';

interface Vector {
  name: string;
  input: {
    method: string;
    baseUrl: string;
    bucket: string;
    key: string;
    secret: string;
    expires: number;
    range?: string;
    responseContentType?: string;
    responseContentDisposition?: string;
  };
  now: number;
  canonical: string;
  url: string;
  verifyRange?: string;
}

const here = dirname(fileURLToPath(import.meta.url));
const vectors = JSON.parse(readFileSync(resolve(here, '../../testdata/edge-signing-vectors.json'), 'utf8')) as Vector[];

test('shared edge signing vectors', () => {
  for (const vector of vectors) {
    const raw = signUrl({
      method: vector.input.method,
      baseUrl: vector.input.baseUrl,
      bucket: vector.input.bucket,
      key: vector.input.key,
      secret: vector.input.secret,
      expires: vector.input.expires,
      range: vector.input.range,
      responseContentType: vector.input.responseContentType,
      responseContentDisposition: vector.input.responseContentDisposition,
    });
    assert.equal(raw, vector.url, vector.name);

    const claims = verifyUrl({
      method: vector.input.method,
      url: raw,
      secret: vector.input.secret,
      now: vector.now,
      range: vector.verifyRange,
    });
    assert.equal(canonicalString(claims), vector.canonical, vector.name);
  }
});

test('range header must be covered by the signature', () => {
  const raw = signUrl({
    method: 'GET',
    baseUrl: 'https://files.example.com',
    bucket: 'demo-bucket',
    key: 'object.txt',
    secret: 'secret',
    expires: 1780938000,
  });
  assert.throws(
    () => verifyUrl({ method: 'GET', url: raw, secret: 'secret', now: 1780934400, range: 'bytes=0-9' }),
    UnsignedRangeError,
  );

  const ranged = signUrl({
    method: 'GET',
    baseUrl: 'https://files.example.com',
    bucket: 'demo-bucket',
    key: 'object.txt',
    secret: 'secret',
    expires: 1780938000,
    range: 'bytes=0-9',
  });
  assert.throws(
    () => verifyUrl({ method: 'GET', url: ranged, secret: 'secret', now: 1780934400, range: 'bytes=1-9' }),
    RangeMismatchError,
  );
});

test('multi-server signing inserts server after base path prefix', () => {
  const raw = signUrl({
    method: 'GET',
    baseUrl: 'https://files.example.com/public/root?existing=1',
    server: 'media_1',
    bucket: 'demo-bucket',
    key: 'dir/object.txt',
    secret: 'secret',
    expires: 1780938000,
  });

  const parsed = new URL(raw);
  assert.equal(parsed.pathname, '/public/root/media_1/demo-bucket/dir/object.txt');
  assert.equal(parsed.searchParams.get('existing'), '1');
});

test('multi-server verification returns server claim and canonical string includes it', () => {
  const raw = signUrl({
    method: 'GET',
    baseUrl: 'https://files.example.com',
    server: 'media-1',
    bucket: 'demo-bucket',
    key: 'dir/object.txt',
    secret: 'secret',
    expires: 1780938000,
  });

  const claims = verifyUrl({ method: 'GET', url: raw, server: 'media-1', secret: 'secret', now: 1780934400 });

  assert.equal(claims.server, 'media-1');
  assert.equal(claims.bucket, 'demo-bucket');
  assert.equal(claims.key, 'dir/object.txt');
  assert.equal(
    canonicalString(claims),
    ['GET', 'media-1', 'demo-bucket', 'dir/object.txt', '1780938000', '', '', ''].join('\n'),
  );
});

test('multi-server verification rejects tampered server path', () => {
  const raw = signUrl({
    method: 'GET',
    baseUrl: 'https://files.example.com',
    server: 'media-1',
    bucket: 'demo-bucket',
    key: 'object.txt',
    secret: 'secret',
    expires: 1780938000,
  });
  const tampered = raw.replace('/media-1/', '/media-2/');

  assert.throws(
    () => verifyUrl({ method: 'GET', url: tampered, server: 'media-2', secret: 'secret', now: 1780934400 }),
    InvalidSignatureError,
  );
});

test('multi-server verification rejects expected-server mismatch', () => {
  const raw = signUrl({
    method: 'GET',
    baseUrl: 'https://files.example.com',
    server: 'media-1',
    bucket: 'demo-bucket',
    key: 'object.txt',
    secret: 'secret',
    expires: 1780938000,
  });

  assert.throws(
    () => verifyUrl({ method: 'GET', url: raw, server: 'media-2', secret: 'secret', now: 1780934400 }),
    InvalidSignatureError,
  );
});

test('default bucket short path signs real bucket and verifies explicitly', () => {
  const raw = signUrl({
    method: 'GET',
    baseUrl: 'https://files.example.com',
    server: 'blue',
    bucket: 'demo-bucket',
    key: 'archive/object.txt',
    secret: 'secret',
    expires: 1780938000,
    defaultBucketPath: true,
  });

  const parsed = new URL(raw);
  assert.equal(parsed.pathname, '/blue/archive/object.txt');

  assert.throws(
    () => verifyUrl({ method: 'GET', url: raw, server: 'blue', secret: 'secret', now: 1780934400 }),
    InvalidSignatureError,
  );

  const claims = verifyUrl({
    method: 'GET',
    url: raw,
    server: 'blue',
    secret: 'secret',
    now: 1780934400,
    defaultBucket: 'demo-bucket',
  });
  assert.equal(claims.server, 'blue');
  assert.equal(claims.bucket, 'demo-bucket');
  assert.equal(claims.key, 'archive/object.txt');
  assert.equal(
    canonicalString(claims),
    ['GET', 'blue', 'demo-bucket', 'archive/object.txt', '1780938000', '', '', ''].join('\n'),
  );
});

test('default bucket path requires server', () => {
  assert.throws(
    () => signUrl({
      method: 'GET',
      baseUrl: 'https://files.example.com',
      bucket: 'demo-bucket',
      key: 'object.txt',
      secret: 'secret',
      expires: 1780938000,
      defaultBucketPath: true,
    }),
    EdgeSignError,
  );
});

test('empty server keeps legacy signing and verification behavior', () => {
  const input = {
    method: 'GET',
    baseUrl: 'https://files.example.com',
    bucket: 'demo-bucket',
    key: 'object.txt',
    secret: 'secret',
    expires: 1780938000,
  };
  const legacy = signUrl(input);
  const explicitEmpty = signUrl({ ...input, server: '' });

  assert.equal(explicitEmpty, legacy);

  const claims = verifyUrl({ method: 'GET', url: explicitEmpty, server: '', secret: 'secret', now: 1780934400 });
  assert.equal(claims.server, undefined);
  assert.equal(canonicalString(claims), ['GET', 'demo-bucket', 'object.txt', '1780938000', '', '', ''].join('\n'));
});
